doc/proposals/scorecard-user-tests.md

```diff
@@ -0,0 +1,112 @@
+# User Defined Tests for the Operator Scorecard
+
+Implementation Owner: AlexNPavel
+
+Status: Draft
+
+[Background](#Background)
+
+[Goals](#Goals)
+
+[Design overview](#Design_overview)
+
+[User facing usage](#User_facing_usage)
+
+[Observations and open questions](#Observations_and_open_questions)
+
+## Background
+
+The operator scorecard is intended to allow users to run a generic set of tests on their operators. Some simple checks can be performed, but more complicated
+tests verifying that the operator actually works are not possible to do in a generic way. This leads to some of the functional tests in the current scorecard
+implementation to be inaccurate or too insufficient to be useful. For more useful functional scorecard tests, we need to allow some basic user input for tests.
+
+## Goals
+
+- Implement user-defined scorecard tests
+- Replace existing "Operator actions are reflected in status" and "Writing into CRs has effects" tests with user-defined tests
+
+## Design overview
+
+A new section would be added to the scorecard config file called `functional_tests` where a user can define various aspects of a test. For example, this definition runs a similar test to the memcached-operator scale test from the SDK's e2e test:
+
+```yaml
+functional_tests:
+  - cr: "deploy/crds/memcached.cr.yaml"
+    expected_resources:
+      - kind: Deployment
+        name: "example_memcached"
+        fields:
+          status:
+            readyReplicas: 3
+          spec:
+            template:
+              spec:
+                containers:
+                  - image: memcached:1.4.36-alpine
+    expected_status:
+      scorecard_function_length:
+        nodes: 3
+    modifications:
+      - spec:
+        - size: 4
+        expected_resources:
+          - kind: Deployment
+            name: "example_memcached"
```

Comment:
Will we need `apiVersion` to be included here as well? For example, what if the described resource is a custom `Deployment` type defined in the `my.custom.org` group?

Also, are we trying to support various scope use cases, e.g. cluster-scoped CR creates a number of objects in various namespaces? If so, we may also need a `namespace` field.

Reply:
Yeah, we want to support all operators/CRs. I'll add an apiVersion and namespace field (we can assume that if namespace is not specified that we just use the current namespace of the CR).